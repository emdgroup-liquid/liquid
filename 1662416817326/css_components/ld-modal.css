:host{display:contents}@keyframes ld-modal-in{0%{transform:translateY(2rem)}to{transform:translateY(0)}}@keyframes ld-modal-out{0%{transform:scale(1)}to{transform:scale(.9)}}@keyframes ld-modal-out-mobile{0%{transform:translateY(0)}to{transform:translateY(2rem)}}:host,dialog.ld-modal{--ld-modal-bg-col:var(--ld-col-wht);--ld-modal-fixed-padding-x:var(--ld-sp-16);--ld-modal-fixed-padding-y:var(--ld-sp-16);--ld-modal-padding-x:var(--ld-sp-16);--ld-modal-padding-y:var(--ld-sp-24);--ld-modal-fixed-bg-col:var(--ld-col-neutral-010);--ld-modal-transition-duration:0.001s;--ld-modal-max-inline-size:30rem;--ld-modal-max-block-size:70rem;--ld-modal-min-inline-size:18rem}@media (prefers-reduced-motion:no-preference){:host,dialog.ld-modal{--ld-modal-transition-duration:0.2s}}:host dialog,dialog.ld-modal{animation:ld-modal-in var(--ld-modal-transition-duration) ease-out forwards;border:0;border-radius:var(--ld-br-l);box-shadow:var(--ld-shadow-active);color:var(--ld-col-neutral-900);display:flex;flex-direction:column;inset:0;margin:auto;max-block-size:min(calc(100% - var(--ld-sp-24) - var(--ld-sp-40)),var(--ld-modal-max-block-size));max-inline-size:min(calc(100% - var(--ld-sp-32)),var(--ld-modal-max-inline-size));min-inline-size:var(--ld-modal-min-inline-size);overflow:visible;padding:0;position:fixed;transition:opacity var(--ld-modal-transition-duration) linear,transform var(--ld-modal-transition-duration) ease;z-index:2147483647}@media (max-width:32rem){:host dialog,dialog.ld-modal{margin-bottom:var(--ld-sp-40)}}:host dialog:not([open]),dialog.ld-modal:not([open]){animation:ld-modal-out var(--ld-modal-transition-duration) ease-in forwards;opacity:0;pointer-events:none;transition:opacity var(--ld-modal-transition-duration) linear,visibility 0s var(--ld-modal-transition-duration) linear,transform var(--ld-modal-transition-duration) ease;visibility:hidden}@media (max-width:32rem){:host dialog:not([open]),dialog.ld-modal:not([open]){animation-name:ld-modal-out-mobile}}:host dialog:after,dialog.ld-modal:after{background-color:var(--ld-modal-bg-col);border-radius:inherit;content:"";inset:0;position:absolute;z-index:-1}:host dialog:before,dialog.ld-modal:before{background-color:var(--ld-thm-primary-active);content:"";inset:0;opacity:.3;position:fixed;transform:scale(99);z-index:-2}:host(.ld-modal--blurry-backdrop) dialog::-webkit-backdrop,dialog.ld-modal--blurry-backdrop::-webkit-backdrop{-webkit-backdrop-filter:blur(5px);backdrop-filter:blur(5px)}:host(.ld-modal--blurry-backdrop) dialog::backdrop,dialog.ld-modal--blurry-backdrop::backdrop{-webkit-backdrop-filter:blur(5px);backdrop-filter:blur(5px)}.ld-modal__x{align-self:flex-start;-webkit-appearance:none;appearance:none;aspect-ratio:1;background-color:initial;border:var(--ld-sp-8) solid #0000;box-sizing:border-box;cursor:pointer;display:inline-flex;justify-self:flex-end;margin:var(--ld-sp-8) calc(var(--ld-sp-8)*-1) var(--ld-sp-8) auto;overflow:hidden;position:relative;width:var(--ld-sp-32)}.ld-modal__x:after,.ld-modal__x:before{background-color:var(--ld-col-neutral-900);border-radius:1rem;content:"";display:block;height:125%;left:50%;position:absolute;top:50%;width:.15rem}.ld-modal__x:before{transform:translateX(-50%) translateY(-50%) rotate(45deg)}.ld-modal__x:after{transform:translateX(-50%) translateY(-50%) rotate(-45deg)}.ld-modal__content{flex-grow:1;overflow-x:hidden;overflow-y:auto;overflow:hidden auto;overscroll-behavior:none;padding:var(--ld-modal-padding-y) var(--ld-modal-padding-x)}.ld-modal__footer,.ld-modal__header{align-items:center;background-color:var(--ld-modal-fixed-bg-col);display:grid;gap:var(--ld-modal-fixed-padding-x);grid-auto-flow:column;padding:0 var(--ld-modal-fixed-padding-x)}.ld-modal__header{border-radius:var(--ld-br-l) var(--ld-br-l) 0 0}.ld-modal__header>::slotted(:not(.ld-modal__x)),.ld-modal__header>:not(.ld-modal__x){padding:var(--ld-sp-12) 0}.ld-modal__footer{border-radius:0 0 var(--ld-br-l) var(--ld-br-l);justify-content:flex-end}.ld-modal__footer>*,.ld-modal__footer>::slotted(*){margin:var(--ld-modal-fixed-padding-y) 0}