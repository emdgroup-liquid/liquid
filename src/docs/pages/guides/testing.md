---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

