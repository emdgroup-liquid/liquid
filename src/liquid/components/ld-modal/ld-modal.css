:host {
  display: contents;
}

@keyframes ld-modal-in {
  from {
    transform: translateY(2rem);
  }

  to {
    transform: translateY(0);
  }
}

@keyframes ld-modal-out {
  from {
    transform: scale(1);
  }

  to {
    transform: scale(0.9);
  }
}

@keyframes ld-modal-out-mobile {
  from {
    transform: translateY(0);
  }

  to {
    transform: translateY(2rem);
  }
}

dialog.ld-modal,
:host {
  --ld-modal-bg-col: var(--ld-col-wht);
  --ld-modal-fixed-padding-x: var(--ld-sp-16);
  --ld-modal-fixed-padding-y: var(--ld-sp-16);
  --ld-modal-padding-x: var(--ld-sp-16);
  --ld-modal-padding-y: var(--ld-sp-24);
  --ld-modal-fixed-bg-col: var(--ld-col-neutral-010);
  --ld-modal-transition-duration: var(--ld-transition-duration-instant);
  --ld-modal-max-inline-size: 30rem;
  --ld-modal-max-block-size: 70rem;
  --ld-modal-min-inline-size: 18rem;

  @media (prefers-reduced-motion: no-preference) {
    --ld-modal-transition-duration: var(--ld-transition-duration-normal);
  }
}

dialog.ld-modal,
:host dialog {
  /* overwrites */
  border: 0;
  color: var(--ld-col-neutral-900);
  color-scheme: var(--ld-modal-color-scheme, none);
  display: flex; /* allows for transitions */
  flex-direction: column;
  inset: 0; /* keeps the dialog positioned correctly during closing transition */
  margin: auto;
  overflow: visible; /* required due to scalehack */
  padding: 0;
  position: fixed; /* makes sure it stays fixed during closing transition */
  z-index: 2147483647; /* makes sure it is not overlapped during closing transition */

  animation: ld-modal-in var(--ld-modal-transition-duration) ease-out forwards;
  border-radius: var(--ld-br-l);
  box-shadow: var(--ld-shadow-active);

  /* dimensions */
  max-block-size: min(
    /* account for additional margin bottom on mobile */
      calc(100% - var(--ld-sp-24) - var(--ld-sp-40)),
    var(--ld-modal-max-block-size)
  );
  max-inline-size: min(
    calc(100% - var(--ld-sp-32)),
    var(--ld-modal-max-inline-size)
  );
  min-inline-size: var(--ld-modal-min-inline-size);

  transition: opacity var(--ld-modal-transition-duration) linear,
    transform var(--ld-modal-transition-duration) ease;

  @media (width <= 32rem) {
    /*
    On mobile we place the dialog at the bottom of the screen
    so that it is easier for the user to interact with it. */
    margin-bottom: var(--ld-sp-40);
  }

  &:not([open]) {
    animation: ld-modal-out var(--ld-modal-transition-duration) ease-in forwards;
    opacity: 0;
    pointer-events: none;
    transition: opacity var(--ld-modal-transition-duration) linear,
      visibility 0s var(--ld-modal-transition-duration) linear,
      transform var(--ld-modal-transition-duration) ease;
    visibility: hidden;

    @media (width <= 32rem) {
      /*
      On mobile, since the dialog is placed at the bottom of the screen,
      we transition it out slightly differently than on wide view ports. */
      animation-name: ld-modal-out-mobile;
    }
  }

  &::after {
    background-color: var(--ld-modal-bg-col);
    border-radius: inherit;
    content: '';
    inset: 0;
    position: absolute;
    z-index: -1;
  }

  &::before {
    background-color: var(--ld-thm-primary-active);
    content: '';
    inset: 0;
    opacity: 0.3;
    position: fixed;
    transform: scale(99); /* scalehack required due to animation */
    z-index: -2;
  }
}

dialog.ld-modal--blurry-backdrop,
:host(.ld-modal--blurry-backdrop) dialog {
  &::backdrop {
    backdrop-filter: blur(5px);
  }
}

/* custom icon cross */
.ld-modal__x {
  place-self: flex-start flex-end;
  appearance: none;
  background-color: transparent;
  border: solid transparent var(--ld-sp-8);
  box-sizing: border-box;
  cursor: pointer;
  display: inline-flex;
  height: 3rem;
  margin: auto calc(var(--ld-sp-16) * -1) auto auto;
  overflow: hidden;
  position: relative;
  width: 3rem;

  &::before,
  &::after {
    border-radius: 1rem;
    background-color: var(--ld-col-neutral-900);
    content: '';
    display: block;
    height: 1.25rem;
    left: 50%;
    position: absolute;
    top: 50%;
    width: 0.15rem;
  }

  &::before {
    transform: translateX(-50%) translateY(-50%) rotate(45deg);
  }

  &::after {
    transform: translateX(-50%) translateY(-50%) rotate(-45deg);
  }
}

.ld-modal__content {
  flex-grow: 1;
  overflow: hidden auto;
  overscroll-behavior: none;
  padding: var(--ld-modal-padding-y) var(--ld-modal-padding-x);
}

.ld-modal__header,
.ld-modal__footer {
  align-items: center;
  background-color: var(--ld-modal-fixed-bg-col);
  display: grid;
  grid-auto-flow: column;
  gap: var(--ld-modal-fixed-padding-x);
  padding: 0 var(--ld-modal-fixed-padding-x);
}

.ld-modal__header {
  border-radius: var(--ld-br-l) var(--ld-br-l) 0 0;

  > *:not(.ld-modal__x),
  > ::slotted(*:not(.ld-modal__x)) {
    padding: var(--ld-sp-12) 0;
  }
}

.ld-modal__footer {
  border-radius: 0 0 var(--ld-br-l) var(--ld-br-l);
  justify-content: flex-end;

  > *,
  > ::slotted(*) {
    margin: var(--ld-modal-fixed-padding-y) 0;
  }
}
