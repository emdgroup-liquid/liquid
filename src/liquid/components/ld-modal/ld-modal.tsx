import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  Host,
  Listen,
  Method,
  Prop,
  Watch,
} from '@stencil/core'
import { getClassNames } from '../../utils/getClassNames'

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part dialog - Actual `dialog` element
 * @part content - `div` element wrapping the default slot
 * @part footer - `footer` element
 * @part header - `header` element
 */
@Component({
  tag: 'ld-modal',
  styleUrl: 'ld-modal.css',
  shadow: true,
})
export class LdModal {
  @Element() el: HTMLElement
  private dialogRef: HTMLDialogElement

  /** The modal is cancelable by default. However, you can change this using this prop. */
  @Prop() cancelable?: boolean = true

  /** Indicates that the modal dialog is active and can be interacted with. */
  @Prop({ mutable: true, reflect: true }) open?: boolean

  /** Use a blurry backdrop. */
  @Prop() blurryBackdrop?: boolean = false

  /** Emitted when modal is opening (before transition). */
  @Event() ldmodalopening: EventEmitter

  /** Emitted when modal has opened (after transition). */
  @Event() ldmodalopened: EventEmitter

  /** Emitted when modal is closing (before transition). */
  @Event() ldmodalclosing: EventEmitter

  /** Emitted when modal has closed (after transition). */
  @Event() ldmodalclosed: EventEmitter

  /** Opens the modal dialog. */
  @Method()
  async showModal() {
    this.open = true
  }

  /** Closes the modal dialog. */
  @Method()
  async close() {
    this.open = false
  }

  @Listen('keydown', { passive: true, target: 'window' })
  handleKeyDown(ev: KeyboardEvent) {
    if (ev.key === 'Escape' && this.cancelable) {
      this.open = false
    }
  }

  @Watch('open')
  onOpenChange(open: boolean) {
    // Calling the showModal and close methods on the dialog element here
    // is super important, because these make the native focus trap and
    // the backdrop feature work.
    // TODO: Remove @ts-ignore comments as soon as TS types get updated.
    if (open) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      this.dialogRef.showModal()
      this.ldmodalopening.emit()
    } else {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      this.dialogRef.close()
      this.ldmodalclosing.emit()
    }
  }

  private handleClose = () => {
    // When the dialog is closed with the Esc key we need to
    // update the open prop explicitly.
    this.open = false
  }

  private handleCancel = (ev: Event) => {
    if (!this.cancelable) {
      ev.preventDefault()
    }
  }

  private handleClick = (ev: MouseEvent) => {
    if (this.cancelable && (ev.target as HTMLElement).tagName === 'DIALOG') {
      this.close()
    }
  }

  private handleTransitionEnd = () => {
    if (this.open) {
      this.ldmodalopened.emit()
    } else {
      this.ldmodalclosed.emit()
    }
  }

  componentDidLoad() {
    this.dialogRef.addEventListener('cancel', this.handleCancel)
  }

  disconnectedCallback() {
    this.dialogRef.removeEventListener('cancel', this.handleCancel)
  }

  render() {
    const cl = getClassNames([
      'ld-modal',
      this.blurryBackdrop && 'ld-modal--blurry-backdrop',
    ])

    return (
      <Host class={cl}>
        <dialog
          onClick={this.handleClick}
          onClose={this.handleClose}
          onTransitionEnd={this.handleTransitionEnd}
          open={this.open}
          part="dialog"
          ref={(el) => (this.dialogRef = el as HTMLDialogElement)}
        >
          <header class="ld-modal__header" part="header">
            <slot name="header"></slot>
            {this.cancelable && (
              <button
                class="ld-modal__x"
                aria-label="Dismiss"
                onClick={this.close.bind(this)}
              ></button>
            )}
          </header>
          <div class="ld-modal__content" part="content">
            <slot></slot>
          </div>
          <footer class="ld-modal__footer" part="footer">
            <slot name="footer"></slot>
          </footer>
        </dialog>
      </Host>
    )
  }
}
