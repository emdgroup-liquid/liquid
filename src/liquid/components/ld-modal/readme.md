---
eleventyNavigation:
  key: Modal
  parent: Components
layout: layout.njk
title: Modal
permalink: components/ld-modal/
tags:
  - dialog
  - alert
  - popup
  - popover
  - lightbox
---

# ld-modal

<link rel="stylesheet" href="css_components/ld-modal.css">
<link rel="stylesheet" href="css_components/ld-typo.css">
<link rel="stylesheet" href="css_components/ld-button.css">

The `ld-modal` component represents a dismissible modal dialog box.

The [`<dialog>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog) element acts as the foundation for both, the Web Component version and the CSS component version of the component. The `ld-modal` Web Component wraps a `<dialog>` element and enhances its functionality which otherwise is not natively implemented, such as by emitting additional events and handling clicks on the dialog backdrop.

## Examples

Here is a minimalistic example of a modal dialog:

{% example %}
<ld-modal>
  <ld-typo style="text-align: center">
    I'm a modal dialog.
  </ld-typo>
</ld-modal>

<ld-button onclick="event.target.previousElementSibling.showModal()">Open Modal</ld-button>

<!-- CSS component -->

<dialog class="ld-modal">
  <header class="ld-modal__header">
    <button
      class="ld-modal__x"
      aria-label="Dismiss"
      onclick="this.closest('dialog').close()"
    ></button>
  </header>
  <div class="ld-modal__content">
    <p class="ld-typo" style="text-align: center">
      I'm a modal dialog.
    </p>
  </div>
</dialog>

<button class="ld-button" onclick="event.target.previousElementSibling.showModal()">Open Modal</button>
{% endexample %}

### With header and footer

You have two additional slots to your disposal for altering the modal header and footer which are both positioned fixed at top and bottom of the dialog element.

{% example %}
<ld-modal>
  <ld-typo slot="header">Hello</ld-typo>
  <ld-typo style="text-align: center">
    I'm a modal dialog.
  </ld-typo>
  <ld-button slot="footer" style="width: 8rem" mode="ghost" onclick="this.closest('ld-modal').close()">Cancel</ld-button>
  <ld-button slot="footer" style="width: 8rem" onclick="this.closest('ld-modal').close()">Submit</ld-button>
</ld-modal>

<ld-button onclick="event.target.previousElementSibling.showModal()">Open Modal</ld-button>

<!-- CSS component -->

<dialog class="ld-modal">
  <header class="ld-modal__header">
    <p class="ld-typo">Hello</p>
    <button
      class="ld-modal__x"
      aria-label="Dismiss"
      onclick="this.closest('dialog').close()"
    ></button>
  </header>
  <div class="ld-modal__content">
    <p class="ld-typo" style="text-align: center">
      I'm a modal dialog.
    </p>
  </div>
  <footer class="ld-modal__footer">
    <button class="ld-button ld-button--ghost" style="width: 8rem" onclick="this.closest('dialog').close()">Cancel</button>
    <button class="ld-button" style="width: 8rem" onclick="this.closest('dialog').close()">Submit</button>
  </footer>
</dialog>

<button class="ld-button" onclick="event.target.previousElementSibling.showModal()">Open Modal</button>
{% endexample %}

### Non-cancelable

{% example %}
<ld-modal cancelable="false">
  <ld-typo slot="header">Hello</ld-typo>
  <ld-typo style="text-align: center">
    I'm a modal dialog.
  </ld-typo>
  <ld-button slot="footer" style="width: 8rem" onclick="this.closest('ld-modal').close()">Submit</ld-button>
</ld-modal>

<ld-button onclick="event.target.previousElementSibling.showModal()">Open Modal</ld-button>

<!-- CSS component -->

<dialog class="ld-modal" oncancel="event.preventDefault()">
  <header class="ld-modal__header">
    <p class="ld-typo">Hello</p>
  </header>
  <div class="ld-modal__content">
    <p class="ld-typo" style="text-align: center">
      I'm a modal dialog.
    </p>
  </div>
  <footer class="ld-modal__footer">
    <button class="ld-button" style="width: 8rem" onclick="this.closest('dialog').close()">Submit</button>
  </footer>
</dialog>

<button class="ld-button" onclick="event.target.previousElementSibling.showModal()">Open Modal</button>
{% endexample %}

### With blurry backdrop

{% example %}
<ld-modal blurry-backdrop>
  <ld-typo style="text-align: center">
    I'm a modal dialog.
  </ld-typo>
</ld-modal>

<ld-button onclick="event.target.previousElementSibling.showModal()">Open Modal</ld-button>

<!-- CSS component -->

<dialog class="ld-modal ld-modal--blurry-backdrop">
  <header class="ld-modal__header">
    <button
      class="ld-modal__x"
      aria-label="Dismiss"
      onclick="this.closest('dialog').close()"
    ></button>
  </header>
  <div class="ld-modal__content">
    <p class="ld-typo" style="text-align: center">
      I'm a modal dialog.
    </p>
  </div>
</dialog>

<button class="ld-button" onclick="event.target.previousElementSibling.showModal()">Open Modal</button>
{% endexample %}

<!-- Auto Generated Below -->


## Properties

| Property         | Attribute         | Description                                                                       | Type               | Default     |
| ---------------- | ----------------- | --------------------------------------------------------------------------------- | ------------------ | ----------- |
| `blurryBackdrop` | `blurry-backdrop` | Use a blurry backdrop.                                                            | `boolean`          | `false`     |
| `cancelable`     | `cancelable`      | The modal is cancelable by default. However, you can change this using this prop. | `boolean`          | `true`      |
| `key`            | `key`             | for tracking the node's identity when working with lists                          | `string \| number` | `undefined` |
| `open`           | `open`            | Indicates that the modal dialog is active and can be interacted with.             | `boolean`          | `undefined` |
| `ref`            | `ref`             | reference to component                                                            | `any`              | `undefined` |


## Events

| Event            | Description                                        | Type               |
| ---------------- | -------------------------------------------------- | ------------------ |
| `ldmodalclosed`  | Emitted when modal has closed (after transition).  | `CustomEvent<any>` |
| `ldmodalclosing` | Emitted when modal is closing (before transition). | `CustomEvent<any>` |
| `ldmodalopened`  | Emitted when modal has opened (after transition).  | `CustomEvent<any>` |
| `ldmodalopening` | Emitted when modal is opening (before transition). | `CustomEvent<any>` |


## Methods

### `close() => Promise<void>`

Closes the modal dialog.

#### Returns

Type: `Promise<void>`



### `showModal() => Promise<void>`

Opens the modal dialog.

#### Returns

Type: `Promise<void>`




## Shadow Parts

| Part        | Description |
| ----------- | ----------- |
| `"content"` |             |
| `"footer"`  |             |
| `"header"`  |             |


----------------------------------------------

 
