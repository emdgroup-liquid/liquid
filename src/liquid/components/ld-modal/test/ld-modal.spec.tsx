import { h } from '@stencil/core'
import { newSpecPage } from '@stencil/core/testing'
import { LdModal } from '../ld-modal'

async function transitionEnd(page) {
  const dialog = page.root.shadowRoot.querySelector('dialog')
  const transitionEndHandler = dialog['__listeners'].find(
    (l) => l.type === 'transitionEnd'
  ).handler
  transitionEndHandler({ target: dialog })
  await page.waitForChanges()
}

function getModal(page) {
  const ldModal = page.root as HTMLLdModalElement
  const dialog = ldModal.shadowRoot.querySelector('dialog')
  dialog['close'] = jest.fn()
  dialog['showModal'] = jest.fn()
  return ldModal
}

describe('ld-modal', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [LdModal],
      template: () => <ld-modal>Hello</ld-modal>,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('is open', async () => {
    const page = await newSpecPage({
      components: [LdModal],
      template: () => <ld-modal open>Hello</ld-modal>,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('uses a blurry backdrop', async () => {
    const page = await newSpecPage({
      components: [LdModal],
      template: () => (
        <ld-modal open blurry-backdrop>
          Hello
        </ld-modal>
      ),
    })
    expect(page.root).toMatchSnapshot()
  })

  it('opens', async () => {
    const page = await newSpecPage({
      components: [LdModal],
      template: () => <ld-modal>Hello</ld-modal>,
    })

    const ldModal = getModal(page)

    ldModal.showModal()
    await page.waitForChanges()

    expect(ldModal.open).toBe(true)
  })

  it('closes via method call', async () => {
    const page = await newSpecPage({
      components: [LdModal],
      template: () => <ld-modal open>Hello</ld-modal>,
    })

    const ldModal = getModal(page)

    ldModal.close()
    await page.waitForChanges()

    expect(ldModal.open).toBe(false)
  })

  it('closes via click on close button', async () => {
    const page = await newSpecPage({
      components: [LdModal],
      template: () => <ld-modal open>Hello</ld-modal>,
    })

    const ldModal = getModal(page)

    ldModal.shadowRoot.querySelector<HTMLButtonElement>('.ld-modal__x').click()
    await page.waitForChanges()

    expect(ldModal.open).toBe(false)
  })

  it('closes via click on backdrop', async () => {
    const page = await newSpecPage({
      components: [LdModal],
      template: () => <ld-modal open>Hello</ld-modal>,
    })

    const ldModal = getModal(page)

    ldModal.shadowRoot.querySelector('dialog').click()
    await page.waitForChanges()

    expect(ldModal.open).toBe(false)
  })

  it('closes via Escape key press', async () => {
    const page = await newSpecPage({
      components: [LdModal],
      template: () => <ld-modal open>Hello</ld-modal>,
    })

    const ldModal = getModal(page)

    window.dispatchEvent(new KeyboardEvent('keydown', { key: 'Escape' }))
    await page.waitForChanges()

    expect(ldModal.open).toBe(false)
  })

  it('does not close via click on dialog content', async () => {
    const page = await newSpecPage({
      components: [LdModal],
      template: () => (
        <ld-modal open>
          <button id="hello">Hello</button>
        </ld-modal>
      ),
    })

    const ldModal = getModal(page)

    ldModal.querySelector<HTMLButtonElement>('#hello').click()
    await page.waitForChanges()

    expect(ldModal.open).toBe(true)
  })

  it('does not display close button in header if not cancelable', async () => {
    const page = await newSpecPage({
      components: [LdModal],
      template: () => (
        <ld-modal open cancelable={false}>
          <span slot="header">Hello</span>
          Hello
        </ld-modal>
      ),
    })
    expect(page.root).toMatchSnapshot()
  })

  it('does not close via click on backdrop if not cancelable', async () => {
    const page = await newSpecPage({
      components: [LdModal],
      template: () => (
        <ld-modal open cancelable={false}>
          Hello
        </ld-modal>
      ),
    })

    const ldModal = getModal(page)

    ldModal.shadowRoot.querySelector('dialog').click()
    await page.waitForChanges()

    expect(ldModal.open).toBe(true)
  })

  it('does not close via Escape key press if not cancelable', async () => {
    const page = await newSpecPage({
      components: [LdModal],
      template: () => (
        <ld-modal open cancelable={false}>
          Hello
        </ld-modal>
      ),
    })

    const ldModal = getModal(page)

    window.dispatchEvent(new KeyboardEvent('keydown', { key: 'Escape' }))
    await page.waitForChanges()

    expect(ldModal.open).toBe(true)
  })

  it('emits ldmodalopening and ldmodalopened events', async () => {
    const page = await newSpecPage({
      components: [LdModal],
      template: () => <ld-modal>Hello</ld-modal>,
    })

    const ldModal = getModal(page)

    const openingHandler = jest.fn()
    const openedHandler = jest.fn()

    ldModal.addEventListener('ldmodalopening', openingHandler)
    ldModal.addEventListener('ldmodalopened', openedHandler)

    expect(openingHandler).toHaveBeenCalledTimes(0)
    expect(openedHandler).toHaveBeenCalledTimes(0)

    ldModal.showModal()
    await page.waitForChanges()

    expect(openingHandler).toHaveBeenCalledTimes(1)
    expect(openedHandler).toHaveBeenCalledTimes(0)

    await transitionEnd(page)

    expect(openingHandler).toHaveBeenCalledTimes(1)
    expect(openedHandler).toHaveBeenCalledTimes(1)
  })

  it('emits ldmodalclosing and ldmodalclosed events', async () => {
    const page = await newSpecPage({
      components: [LdModal],
      template: () => <ld-modal open>Hello</ld-modal>,
    })

    const ldModal = getModal(page)

    const closingHandler = jest.fn()
    const closedHandler = jest.fn()

    ldModal.addEventListener('ldmodalclosing', closingHandler)
    ldModal.addEventListener('ldmodalclosed', closedHandler)

    expect(closingHandler).toHaveBeenCalledTimes(0)
    expect(closedHandler).toHaveBeenCalledTimes(0)

    ldModal.close()
    await page.waitForChanges()

    expect(closingHandler).toHaveBeenCalledTimes(1)
    expect(closedHandler).toHaveBeenCalledTimes(0)

    await transitionEnd(page)

    expect(closingHandler).toHaveBeenCalledTimes(1)
    expect(closedHandler).toHaveBeenCalledTimes(1)
  })

  it('does not prevent cancel by default', () => {
    const ldModal = new LdModal()
    const ev = new Event('cancel')
    ldModal['handleCancel'](ev)
    expect(ev.defaultPrevented).toBeFalsy()
  })

  it('prevents cancel if not cancelable', () => {
    const ldModal = new LdModal()
    ldModal.cancelable = false
    const ev = new Event('cancel')
    ldModal['handleCancel'](ev)
    expect(ev.defaultPrevented).toBeTruthy()
  })

  it('closes on Esc key press', () => {
    const ldModal = new LdModal()
    ldModal.open = true
    ldModal['handleClose']()
    expect(ldModal.open).toBeFalsy()
  })

  it('does not throw when disconnecting before hydration', () => {
    const component = new LdModal()
    component.disconnectedCallback()
  })
})
