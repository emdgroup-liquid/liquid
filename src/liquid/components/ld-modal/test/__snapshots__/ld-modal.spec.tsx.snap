// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-modal does not display close button in header if not cancelable 1`] = `
<ld-modal class="ld-modal" open="">
  <mock:shadow-root>
    <dialog open="" part="dialog">
      <header class="ld-modal__header" part="header">
        <slot name="header"></slot>
      </header>
      <div class="ld-modal__content" part="content">
        <slot></slot>
      </div>
      <footer class="ld-modal__footer" part="footer">
        <slot name="footer"></slot>
      </footer>
    </dialog>
  </mock:shadow-root>
  <span slot="header">
    Hello
  </span>
  Hello
</ld-modal>
`;

exports[`ld-modal is open 1`] = `
<ld-modal class="ld-modal" open="">
  <mock:shadow-root>
    <dialog open="" part="dialog">
      <header class="ld-modal__header" part="header">
        <slot name="header"></slot>
        <button aria-label="Dismiss" class="ld-modal__x"></button>
      </header>
      <div class="ld-modal__content" part="content">
        <slot></slot>
      </div>
      <footer class="ld-modal__footer" part="footer">
        <slot name="footer"></slot>
      </footer>
    </dialog>
  </mock:shadow-root>
  Hello
</ld-modal>
`;

exports[`ld-modal renders 1`] = `
<ld-modal class="ld-modal">
  <mock:shadow-root>
    <dialog part="dialog">
      <header class="ld-modal__header" part="header">
        <slot name="header"></slot>
        <button aria-label="Dismiss" class="ld-modal__x"></button>
      </header>
      <div class="ld-modal__content" part="content">
        <slot></slot>
      </div>
      <footer class="ld-modal__footer" part="footer">
        <slot name="footer"></slot>
      </footer>
    </dialog>
  </mock:shadow-root>
  Hello
</ld-modal>
`;

exports[`ld-modal uses a blurry backdrop 1`] = `
<ld-modal blurry-backdrop="" class="ld-modal ld-modal--blurry-backdrop" open="">
  <mock:shadow-root>
    <dialog open="" part="dialog">
      <header class="ld-modal__header" part="header">
        <slot name="header"></slot>
        <button aria-label="Dismiss" class="ld-modal__x"></button>
      </header>
      <div class="ld-modal__content" part="content">
        <slot></slot>
      </div>
      <footer class="ld-modal__footer" part="footer">
        <slot name="footer"></slot>
      </footer>
    </dialog>
  </mock:shadow-root>
  Hello
</ld-modal>
`;
